src/Datadog.Trace/Span.cs

```diff
@@ -219,8 +221,36 @@ public Span SetTag(string key, string value)
 
                     break;
                 default:
-                    // if not a special tag, just add it to the tag bag
-                    Tags[key] = value;
+                    if (value == null)
+                    {
+                        if (Tags != null)
+                        {
+                            // lock when modifying the collection
+                            lock (_tagsLock)
+                            {
+                                // Agent doesn't accept null tag values,
+                                // remove them instead
+                                Tags.Remove(key);
```

Comment:
This behavior is unchanged, but if `key` is null, this operation will throw an `ArgumentNullException`. Should we be raising this exception earlier instead of in the Dictionary call?

Reply:
I like this idea. We also need to fix a few issues with the logic when `value` is `null`. For example. `SetTag(ManualKeep, null)` should probably set `Context.TraceContext.SamplingPriority = null`, but we're not doing that right now. Are you ok if I make these changes in a separate PR?

Reply:
Yeah that's fine with me, I don't want to block the PR on this